File: Classes/SetClockView.m
//
//  SetClockView.m
//  KidsTimeFun
//
//  Created by Jagmeet Chawla on 4/19/09.
//  Modified by SVV Satyanarayana, under contract to NSC Partners LLC on 28/04/10.
//  Copyright 2009-2013 NSC Partners LLC. All rights reserved.
//

#import "SetClockView.h"
#import "KidsTimeFunAppState.h"

@implementation SetClockView

@synthesize hours;
@synthesize minutes;
@synthesize showMinutesOffsetInHoursHand;

//Macros to convert hours, minutes and seconds to angles in radians
#define hoursToRadiansHM(h, m) ((M_PI/2)-(((h*2*M_PI)/12)+((m/60)*(0.75*2*M_PI/12))))
#define hoursToRadiansH(h) ((M_PI/2)-((h*2*M_PI)/12))
#define minutesToRadians(m) ((M_PI/2)-((m*2*M_PI)/60))
#define secondsToRadians(s) ((M_PI/2)-((s*2*M_PI)/60))

#define radiansToHours(theta_h) ((theta_h*6)/M_PI)
#define radiansToMinutes(theta_m) ((theta_m*30)/M_PI)
//Some convenience macros
#define smallerOf(a, b) a<b?a:b
#define greaterOf(a, b) a>b?a:b

//-(float)hours
//{
//    return round(hours);
//}


- (id)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        // Initialization code
		hourHandFlag = NO;		
		minuteHandFlag = NO;
		firstPass = YES;
		hours = 12;
		minutes = 30;
    }
    return self;
}

- (void)drawRect:(CGRect)rect {
    // Clock Drawing code
	CGContextRef context = UIGraphicsGetCurrentContext ();	// Get current graphics context
	//initialize radial coordinates
	//float r = 0.0f, theta = 0.0f;
	//initialize cartesian coordinates
	float x_0 = 0.0f, y_0 = 0.0f;	//Origin
	float x = 0.0f, y = 0.0f;		//coordinates
	//float x_c = 0.0f, y_c = 0.0f;	//center
	float x_l = 0.0f, y_w = 0.0f;	//bounds
	//Load Origin points
	x_0 = 0.0f;
	y_0 = 0.0f;
	//Load bounds - both length and width
	x_l = self.bounds.size.width;
	y_w = self.bounds.size.height;
	//diameter for the clock should be smaller of the two dimensions
	int d = 0.0f;	//initialize diameter
	d = smallerOf(x_l-x_0, y_w-y_0);	//load diameter
	r = d/2;	//radius = diameter/2
	//So the center for this circle (clock) should be
	x_c = (x_0+x_l)/2;
	y_c = (y_0+y_w)/2;
	//and clock face image start points will be
	x = x_c-r;
	y = y_c-r;
	//Load Clock Face with the start points and diameter above (circle in a square)
	UIImage *img = [UIImage imageNamed:@"ClockFace218.png"]; //Load clock face image
	//Draw clock face - with diameter
	[img drawInRect:CGRectMake(x, y, d, d)];
	
	//initialize hours to 12 and minutes to 15
	//self.hours = 12;
	//self.minutes = 30;
	self.showMinutesOffsetInHoursHand = NO;
	
	//Now draw hours hand
	//Get radial coordinates
	//float r_h = 0.0f;	//initialize radius for the hand
	r_h = r*0.65;	//65% of the clock face radius - it is smallest hand
	//Now load angle based on the hours
	if (showMinutesOffsetInHoursHand)
		theta_h = hoursToRadiansHM(self.hours, self.minutes);
	else
		theta_h = hoursToRadiansH(self.hours);
	
	//Convert to cartesian - remember in CorGraphics y coordinate is reversed (-y)
	//And also it's origin will be the center of clock i.e. x_c, y_c
	x = r_h * cos(theta_h);
	y = -1 * r_h * sin(theta_h);
	//now convert these to coordinates from origin x_0, y_0
	x = x+x_c;
	y = y+y_c;
		
	if (hourHandFlag && !firstPass) {
		theta = atan2(x_c-xx,yy-y_c)+M_PI/2;	//This gets the phasing correct
		x = r*0.65 * cos(theta)+x_c;
		y = r*0.65 * sin(theta)+y_c;
		float correctedHourTheata = theta+(M_PI/2);
		self.hours = radiansToHours(correctedHourTheata);
	}
	
		//NSLog(@"%f %f %f %f",x_c,y_c,x,y);
	//Now draw a line 3 pixels wide from center (x_c, y_c) to the above point (x, y)
	CGContextSetLineWidth (context, 7.0f);
	CGContextSetLineCap(context, kCGLineCapRound);
	CGContextSetAlpha(context, 1.0f);
	CGContextSetStrokeColorWithColor (context, [UIColor blueColor].CGColor);
	CGContextSetFillColorWithColor (context, [UIColor blueColor].CGColor);
	CGContextSetShadow(context, CGSizeMake(1.0f, 1.0f), 1.0f);
	CGContextMoveToPoint (context, x_c, y_c);
	xHourHand = x;
	yHourHand = y;
	CGContextAddLineToPoint (context, x, y);
	//Stroke Path Now
	CGContextStrokePath (context);
	/*
	//now draw a circle here with the radius of 20 pixels for dragging
	CGContextSetLineWidth (context, 1.0f);
	CGContextSetLineCap(context, kCGLineCapRound);
	CGContextSetAlpha(context, 0.50f);
	CGContextSetStrokeColorWithColor (context, [UIColor blueColor].CGColor);
	CGContextSetFillColorWithColor (context, [UIColor blueColor].CGColor);	
	CGContextFillEllipseInRect(context, CGRectMake(x-20, y-20, 40, 40));
	CGContextSetLineWidth (context, 1.0f);
	CGContextSetLineCap(context, kCGLineCapRound);
	CGContextSetAlpha(context, 0.75f);
	CGContextSetStrokeColorWithColor (context, [UIColor blueColor].CGColor);
	CGContextAddEllipseInRect(context, CGRectMake(x-20, y-20, 40, 40));
	CGContextStrokePath (context);	
	*/
	//Now minutes hand
	//Get radial coordinates
	//float r_m = 0.0f;
	r_m = r*0.80;
	theta_m = minutesToRadians(self.minutes);
	//Convert to cartesian - remember in CorGraphics y coordinate is reversed (-y)
	//And also it's origin will be the center of clock i.e. x_c, y_c
	x = r_m * cos(theta_m);
	y = -1 * r_m * sin(theta_m);
	//now convert these to coordinates from origin x_0, y_0
	x = x+x_c;
	y = y+y_c;
	
	if (minuteHandFlag && !firstPass) {
		theta = atan2(x_c-xx,yy-y_c)+M_PI/2;	//This gets the phasing correct
		x = r*0.85 * cos(theta)+x_c;
		y = r*0.85 * sin(theta)+y_c;
		float correctedMinutesTheta = theta + (M_PI/2);
		self.minutes = radiansToMinutes(correctedMinutesTheta);
		self.minutes = (self.minutes == 60.0 ? 59.99 : self.minutes);
	}
	
	//Krishna added the following two lines
	//self.hours = round(self.hours);
    self.minutes = (round(self.minutes) == 60 ? 00 : round(self.minutes));
	
		//NSLog(@"Time is %f:%f",self.hours,self.minutes);
	//Now draw a line 3 pixels wide from center (x_c, y_c) to the above point (x, y)
	CGContextSetLineWidth (context, 5.0f);
	CGContextSetLineCap(context, kCGLineCapRound);
	CGContextSetAlpha(context, 1.0f);
	CGContextSetStrokeColorWithColor (context, [UIColor greenColor].CGColor);
	CGContextSetFillColorWithColor (context, [UIColor greenColor].CGColor);
	CGContextSetShadow(context, CGSizeMake(1.0f, 1.0f), 1.0f);
	CGContextMoveToPoint (context, x_c, y_c);
	xMinuteHand = x;
	yMinuteHand = y;
	CGContextAddLineToPoint (context, x, y);
	//Stroke Path Now
	CGContextStrokePath (context);
	//now draw a circle here with the radius of 20 pixels for dragging
	/*
	CGContextSetLineWidth (context, 1.0f);
	CGContextSetLineCap(context, kCGLineCapRound);
	CGContextSetAlpha(context, 0.50f);
	CGContextSetStrokeColorWithColor (context, [UIColor greenColor].CGColor);
	CGContextSetFillColorWithColor (context, [UIColor greenColor].CGColor);	
	CGContextFillEllipseInRect(context, CGRectMake(x-20, y-20, 40, 40));
	CGContextSetLineWidth (context, 1.0f);
	CGContextSetLineCap(context, kCGLineCapRound);
	CGContextSetAlpha(context, 0.75f);
	CGContextSetStrokeColorWithColor (context, [UIColor greenColor].CGColor);
	CGContextAddEllipseInRect(context, CGRectMake(x-20, y-20, 40, 40));
	CGContextStrokePath (context);
	*/
	//Now draw a circle 8 pixels radius from center (x_c, y_c)
	CGContextSetLineWidth (context, 2.0f);
	CGContextSetLineCap(context, kCGLineCapRound);
	CGContextSetAlpha(context, 1.0f);
	CGContextSetStrokeColorWithColor (context, [UIColor blackColor].CGColor);
	CGContextSetFillColorWithColor (context, [UIColor blackColor].CGColor);
	CGContextMoveToPoint (context, x_c, y_c);
	CGContextAddEllipseInRect(context, CGRectMake(x_c-4, y_c-4, 8, 8));
	CGContextFillEllipseInRect(context, CGRectMake(x_c-4, y_c-4, 8, 8));
	CGContextSetShadow(context, CGSizeMake(1.0f, 1.0f), 1.0f);
	//Stroke Path Now
	CGContextStrokePath (context);
	
	firstPass = NO;
    
    if (self.hours == 12.0) {
        self.hours = 0;
    }
    NSLog(@"%f",self.hours);

}


- (void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event {
	
	// touches is an NSSet.  Take any single UITouch from the set
	UITouch *touch = [touches anyObject];

	xx = [touch locationInView:self].x;		//Get touch location
	yy = [touch locationInView:self].y;
	
	//r_m = sqrt(xx * xx + yy * yy);		//Used only for testing

//	xx = r*0.60 * cos(theta)+x_c;
//	yy = r*0.60 * sin(theta)+y_c;
//NSLog(@"TOUCH Moved");
	
	[self setNeedsDisplay];
}
	
- (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event {

		hourHandFlag = NO;
		minuteHandFlag = NO;
	//NSLog(@"TOUCH ENDED");

}

- (void) touchesBegan:(NSSet*)touches withEvent:(UIEvent*)event {
	
	UITouch *touch = [touches anyObject];
	
	xx = [touch locationInView:self].x;		//Get touch location
	yy = [touch locationInView:self].y;	
	NSLog(@"TOUCH Began");
	thetaTouch = atan2(yy-y_c, xx-x_c);
	thetaHoursHand = atan2(yHourHand-y_c,xHourHand-x_c);
	thetaMinutesHand = atan2(yMinuteHand-y_c,xMinuteHand-x_c);
	NSLog(@"Touch %2.2f, Hours %2.2f, Minutes %2.2f",thetaTouch,thetaHoursHand,thetaMinutesHand);
	if (fabs(thetaTouch-thetaHoursHand) < fabs(thetaTouch -thetaMinutesHand))   {
		NSLog(@"Closer to Hours Hand");
		hourHandFlag = YES;
		minuteHandFlag = NO;
	}
	else {
		NSLog(@"Closer to Minutes Hand");
		hourHandFlag = NO;
		minuteHandFlag = YES;
	}

	
}


- (void)dealloc {
    [super dealloc];
}


@end

File: Classes/SetClockView.swift
// Revised by Krishna Narayan on 5/30/26 — Used Claude to migrate to Swift, fix UI Views, remove deprecations, update for iPad, modernize for Apple UI rules.
// Revised by Krishna Narayan on 6/3/26 — Using Claude changed to 1st, 2nd, and 3rd grade levels, belts are earned not selected, added adaptive weak-drilling algorithm to rectify mistakes and build proficiency after initially providing randomized problems for activities
// Copyright 2026 Island Innovation LLC.  All rights reserved.

import UIKit

@objc(SetClockView)
class SetClockView: UIView {

    var hours: Float = 12
    var minutes: Float = 30
    var showMinutesOffsetInHoursHand: Bool = false

    private var r: CGFloat = 0, xc: CGFloat = 0, yc: CGFloat = 0
    private var rh: CGFloat = 0, rm: CGFloat = 0
    private var thetaH: CGFloat = 0, thetaM: CGFloat = 0
    private var xHourHand: CGFloat = 0, yHourHand: CGFloat = 0
    private var xMinuteHand: CGFloat = 0, yMinuteHand: CGFloat = 0
    private var touchX: CGFloat = 0, touchY: CGFloat = 0
    private var hourHandFlag = false, minuteHandFlag = false, firstPass = true

    // VoiceOver: describe the clock and let the user drag the hands directly
    // (allowsDirectInteraction passes touches straight through to this view).
    override var isAccessibilityElement: Bool {
        get { true }
        set { }
    }
    override var accessibilityLabel: String? {
        get {
            let now = ktfSpokenTime(hours: Int(hours.rounded()), minutes: Int(minutes.rounded()))
            return "Adjustable clock, currently set to \(now). Touch and drag the hour and minute hands to set the time."
        }
        set { }
    }
    override var accessibilityTraits: UIAccessibilityTraits {
        get { [.allowsDirectInteraction] }
        set { }
    }

    override func draw(_ rect: CGRect) {
        guard let context = UIGraphicsGetCurrentContext() else { return }

        let x0: CGFloat = 0, y0: CGFloat = 0
        let xl = bounds.size.width
        let yw = bounds.size.height
        let d = min(xl - x0, yw - y0)
        r = d / 2
        xc = (x0 + xl) / 2
        yc = (y0 + yw) / 2

        UIImage(named: "Clock Face")?.draw(in: CGRect(x: xc - r, y: yc - r, width: d, height: d))

        showMinutesOffsetInHoursHand = false

        // Hours hand
        rh = r * 0.65
        if showMinutesOffsetInHoursHand {
            thetaH = (CGFloat.pi / 2) - (((CGFloat(hours) * 2 * CGFloat.pi) / 12) + ((CGFloat(minutes) / 60) * (0.75 * 2 * CGFloat.pi / 12)))
        } else {
            thetaH = (CGFloat.pi / 2) - ((CGFloat(hours) * 2 * CGFloat.pi) / 12)
        }

        var hx = rh * cos(thetaH) + xc
        var hy = -rh * sin(thetaH) + yc

        if hourHandFlag && !firstPass {
            let theta = atan2(xc - touchX, touchY - yc) + CGFloat.pi / 2
            hx = r * 0.65 * cos(theta) + xc
            hy = r * 0.65 * sin(theta) + yc
            hours = Float((theta + CGFloat.pi / 2) * 6 / CGFloat.pi)
        }

        context.setLineWidth(7.0)
        context.setLineCap(.round)
        context.setAlpha(1.0)
        context.setStrokeColor(UIColor.blue.cgColor)
        context.setShadow(offset: CGSize(width: 1, height: 1), blur: 1)
        context.move(to: CGPoint(x: xc, y: yc))
        context.addLine(to: CGPoint(x: hx, y: hy))
        context.strokePath()
        xHourHand = hx
        yHourHand = hy

        // Minutes hand
        rm = r * 0.80
        thetaM = (CGFloat.pi / 2) - ((CGFloat(minutes) * 2 * CGFloat.pi) / 60)
        var mx = rm * cos(thetaM) + xc
        var my = -rm * sin(thetaM) + yc

        if minuteHandFlag && !firstPass {
            let theta = atan2(xc - touchX, touchY - yc) + CGFloat.pi / 2
            mx = r * 0.85 * cos(theta) + xc
            my = r * 0.85 * sin(theta) + yc
            var mins = Float((theta + CGFloat.pi / 2) * 30 / CGFloat.pi)
            if mins == 60 { mins = 59.99 }
            minutes = mins
        }

        minutes = Float(round(Double(minutes)) == 60 ? 0 : round(Double(minutes)))

        context.setLineWidth(5.0)
        context.setLineCap(.round)
        context.setAlpha(1.0)
        context.setStrokeColor(UIColor.green.cgColor)
        context.setShadow(offset: CGSize(width: 1, height: 1), blur: 1)
        context.move(to: CGPoint(x: xc, y: yc))
        context.addLine(to: CGPoint(x: mx, y: my))
        context.strokePath()
        xMinuteHand = mx
        yMinuteHand = my

        // Center circle
        context.setLineWidth(2.0)
        context.setAlpha(1.0)
        context.setStrokeColor(UIColor.black.cgColor)
        context.setFillColor(UIColor.black.cgColor)
        context.fillEllipse(in: CGRect(x: xc - 4, y: yc - 4, width: 8, height: 8))
        context.setShadow(offset: CGSize(width: 1, height: 1), blur: 1)
        context.strokePath()

        firstPass = false

        if hours == 12.0 { hours = 0 }
    }

    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        guard let touch = touches.first else { return }
        touchX = touch.location(in: self).x
        touchY = touch.location(in: self).y
        setNeedsDisplay()
    }

    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        hourHandFlag = false
        minuteHandFlag = false
    }

    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        guard let touch = touches.first else { return }
        touchX = touch.location(in: self).x
        touchY = touch.location(in: self).y

        let thetaTouch = atan2(touchY - yc, touchX - xc)
        let thetaHoursHand = atan2(yHourHand - yc, xHourHand - xc)
        let thetaMinutesHand = atan2(yMinuteHand - yc, xMinuteHand - xc)

        if abs(thetaTouch - thetaHoursHand) < abs(thetaTouch - thetaMinutesHand) {
            hourHandFlag = true
            minuteHandFlag = false
        } else {
            hourHandFlag = false
            minuteHandFlag = true
        }
    }
}
